src/main/scala/com/gu/kindlegen/BookSection.scala

```
@@ -1,17 +1,40 @@
 package com.gu.kindlegen
 
+import scala.collection.breakOut
+
 /**
- * Each Book (eg Guardian or Observer) contains many sections (eg G2, Top Stories, Finance) each of which will have a number of Pages. Each Page can then contain one or more articles
+ * Each Book (eg Guardian or Observer) contains many sections (eg G2, Top Stories, Finance)
  */
-case class BookSection(bookSectionId: String, bookSectionTitle: String, pages: List[BookSectionPage]) {
-  def minPage = pages.map(_.pageNum).min
-  def maxPage = pages.map(_.pageNum).max
-  def numPages = pages.size
-  def numArticles = pages.flatMap(_.articles).size
+case class BookSection(section: Section, articles: Seq[Article]) extends Linkable {
+  private val pageNumbers: Seq[Int] = articles.map(_.newspaperPageNumber)
+  val firstPageNumber: Int = pageNumbers.min
+  val lastPageNumber: Int = pageNumbers.max
+
+  def id: String = section.id
+  def title: String = section.title
+  def link: Link = section.link
+
+  def publicationDate = articles.map(_.pubDate).minBy(_.dateTime)
 }
 
 object BookSection {
+  /** Groups articles into book sections, sorted according to each section's articles' page number */
+  def fromArticles(articles: Seq[Article]): Seq[BookSection] = {
+    articles.groupBy(_.section)
+      .values.map(apply)(breakOut)
+      .sortBy(section => (section.firstPageNumber, section.lastPageNumber, section.title))
+  }
+
+  /** Creates a book section from a collection of articles that all belong to the same section */
+  def apply(articles: Seq[Article]): BookSection = {
+    require(articles.nonEmpty, "A book section must have at least one article!")
```

I'm not entirely opposed to the use of `require`, I can see why it's convenient, particularly in a lambda. It does worry me a bit to see functions that can throw exceptions rather than returning e.g. Try or Either, as it relies on the caller to know this.
Don't let this delay getting the program working, but it's something to consider later on perhaps

---

Thanks. I don't think `Try` and `Either` are very useful, unless we resort to using suppressed exceptions or `Either[T, Seq[Throwable]]`. I honestly think that Scala should add something like Cats' `Validated` to the standard library. Otherwise, we'll keep having to [reinvent the wheel](https://github.com/guardian/kindle-gen/blob/a16decdeffd40694da3e6a4a82543a58298b624f/src/main/scala/com/gu/xml/validation/XmlValidationResult.scala).

On the other hand, I agree with you. Relying on the caller to know that exceptions may be thrown is an undesired pattern. I agree that I should consider it later.

---

I guess each covers different, albeit overlapping, semantic domains. Exceptions (`Try`) should only ever be used for things that are not under the application's control. `Either` covers so many other uses cases (as a bifunctor, for instance), the right-biased aspect is just convention coming from Haskell.

---

I think that the right-biased nature is because it's easier to remember that _"`Right` is right"_ :)

---

If you're just looking to log error strings then `Either[String,T]` is fine here, no?

---

It probably is. Let's revisit it when we've got a bit more time.